torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1320,13 +1320,41 @@ struct to_ir {
     pushFrame(comprehension_block);
     WithInsertPoint guard(comprehension_block);
     auto emit_body = [&]() {
-      auto comprehension_out = emitExpr(lc.elt());
+      Value* out = emitExpr(lc.elt());
+
+      // Make sure that any element types are subtypes of the annotated
+      // type
+      if (type_hint && !out->type()->isSubtypeOf(out->type())) {
+        throw ErrorReport(loc) << "List type annotation did not match "
+                               << "the types of the actual list items";
+      }
+
+      // If we didn't have a type annotation, let the type begin as the
+      // first element in the list
       if (!type_set) {
-        list_value->setType(ListType::create(comprehension_out->type()));
+        list_value->setType(ListType::create(out->type()));
         type_set = true;
       }
+
+      ListTypePtr lt = list_value->type()->expect<ListType>();
+
+      // Is the current element type the same type or a supertype of the
+      // old element type?
+      bool use_new = lt->getElementType()->isSubtypeOf(out->type());
+      // Is the old element type the same type or a supertype of the
+      // current element type?
+      bool use_old = out->type()->isSubtypeOf(lt->getElementType());
+
+      // If we have `use_old && use_new`, we just keep the old element
+      // type. Otherwise...
```

Comment:
Is this comment incomplete?

Reply:
Thanks for pointing out that this comment is unclear. The "Otherwise..." meant "Otherwise, we do the logic directly below this comment". I'll change it!